In [93]:
import pandas as pd
import numpy as np
#imported 2 libraries called the numpy and the pandas

This is the data being used for the peer graded assessment.
The dataset revolves around the compressive strength of various concrete samples, depending on the quantities of diverse components employed in their creation. These components encompass:

1) Cement

2) Blast Furnace Slag

3) Fly Ash

4) Water

5) Superplasticizer

6) Coarse Aggregate

7) Fine Aggregate

In [94]:
#Reading the data through the data dataframe 
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()  #printing the first five rows of the dataset.

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So here we know that the first concrete sample has 540 cubic meter of cement, zero blast furnace slag, zero fly ash, 162 cubic meter of water,2.5 cubic meter of superplatsicizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate, 28 is the age and 79.99 is the strength.

In [95]:
#Checking the total number of data points 
concrete_data.shape

(1030, 9)

In [96]:
#Checking data for any missing values 
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [97]:
# Removing all null data from the dataset 
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

As the data has no null values so now we can use this data to train our model.

For creating a Regression or a classification model we always need to divide the columns into two called the predictors and the target.

In [98]:
# Our target variable is Strength and others columns are the predictors
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # Every columns except Strength
target = concrete_data['Strength']  # Strength column

Printing the predictors and the targets just to check the output

In [99]:
target.head()  #As we have only 1 target variable so one column gets printed

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [100]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

After printing the predictors and the target lets normalize the version of the data by subtracting the mean and diving by the standard deviation

In [101]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [102]:
#Checking the number of predictors 
num_cols = predictors_norm.shape[1]
num_cols

8

Moving forward we import keras and start building our regression model

In [103]:
import keras     #Keras is always imported from the backend tensorflow

In [104]:
from keras.models import Sequential    #We are importing the sequential layer from keras which helps us to create a model layer by layer in a linear stack
from keras.layers import Dense    #A dense layer is imported from layers to form a fully connected neural network layer

Creating the regression model with 3 hidden layers each of 10 nodes and ReLU activation function

In [105]:
def regression_model():    #Created a function to create a regression model.
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(num_cols,)))   #1st hideen layer
    model.add(Dense(10, activation='relu'))                            #2nd hidden layer
    model.add(Dense(10, activation='relu'))                            #3rd hidden layer
    model.add(Dense(1))                                                #Output layer

    model.compile(optimizer='adam', loss='mean_squared_error')  #Compiling the model with the optimizer adam and keeping the loss as MSE.
    return model

Now we want to split the data so we will have to import sklearn to train and test the data

In [106]:
from sklearn.model_selection import train_test_split

Splitting the data into 70 % training and 30% testing data

In [107]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

Now we have to build our model so we call the regression function here that we created above

In [108]:
#Calling our model here
model = regression_model()


In [109]:
#Training our model with epochs=100
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
721/721 [==============================] - 0s 508us/step - loss: 1613.7178
Epoch 2/50
721/721 [==============================] - 0s 116us/step - loss: 1594.6563
Epoch 3/50
721/721 [==============================] - 0s 131us/step - loss: 1580.4468
Epoch 4/50
721/721 [==============================] - 0s 118us/step - loss: 1567.9943
Epoch 5/50
721/721 [==============================] - 0s 132us/step - loss: 1552.2575
Epoch 6/50
721/721 [==============================] - 0s 90us/step - loss: 1528.1794
Epoch 7/50
721/721 [==============================] - 0s 103us/step - loss: 1488.8117
Epoch 8/50
721/721 [==============================] - 0s 96us/step - loss: 1426.4275
Epoch 9/50
721/721 [==============================] - 0s 72us/step - loss: 1333.1939
Epoch 10/50
721/721 [==============================] - 0s 82us/step - loss: 1196.5043
Epoch 11/50
721/721 [==============================] - 0s 80us/step - loss: 1000.3636
Epoch 12/50
721/721 [==============================] - 0s

Now we evaluate the metrics for the model formed 

In [110]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

309/309 [==============================] - 0s 257us/step


135.79331229496927

As per the next step we need to compute the mean squared error

In [111]:
from sklearn.metrics import mean_squared_error

In [112]:
#Printing the mean and standard deviation of the test data 
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

135.79331711894883 0.0


Creating a list of 50 mean squared errors.

In the code below it iterates through a specified number of training and evaluation cycles (total_mean_squared_errors). In each cycle, the dataset is split into training and testing sets, and a neural network model is trained for a fixed number of epochs. The mean squared error (MSE) is then calculated for the model's predictions on the test set, and both the MSE and the mean squared error between true and predicted values are recorded. After the specified number of cycles, the code calculates and prints the mean and standard deviation of the collected mean squared errors, providing insights into the model's overall performance and its variability across multiple runs.

In [113]:
total_mean_squared_errors = 50
#The total number of times the model will be trained and evaluated.
epochs = 50
mean_squared_errors = []
#An empty list to store the mean squared errors for each run.


for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

#Convert the list of mean squared errors to a NumPy array for easier calculation.
#Calculate the mean and standard deviation of the mean squared errors.
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Printing the  mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 101.02310160911584
MSE 2: 112.41851633806445
MSE 3: 57.96284836704291
MSE 4: 47.91496608944001
MSE 5: 42.796963182467856
MSE 6: 41.046390372958385
MSE 7: 42.49708460835577
MSE 8: 33.95855003653221
MSE 9: 33.93680284864308
MSE 10: 31.980814751683702
MSE 11: 32.29058346547741
MSE 12: 29.33970455675835
MSE 13: 33.04645432777775
MSE 14: 35.759313682136415
MSE 15: 30.178591138722442
MSE 16: 23.57092418485475
MSE 17: 27.450808046705127
MSE 18: 29.41396591655645
MSE 19: 27.514729002918624
MSE 20: 32.554340023053115
MSE 21: 30.153306201823707
MSE 22: 27.523531299578718
MSE 23: 25.618781864450202
MSE 24: 24.33652616087287
MSE 25: 30.830211596195753
MSE 26: 29.644175594292797
MSE 27: 24.348525445438124
MSE 28: 24.905116244812998
MSE 29: 32.43675920801255
MSE 30: 26.53937725573296
MSE 31: 25.51015360069892
MSE 32: 24.883319931894444
MSE 33: 23.51155436617657
MSE 34: 29.9054499505793
MSE 35: 32.31971540728819
MSE 36: 29.05329772801075
MSE 37: 22.827059332221072
MSE 38: 26.97895438925734
MSE

With 100 epochs, the mean MSE has increased slightly from 39.44 to 41.03. This means that the additional training epochs might be causing the model to fit the training data more closely, potentially leading to a slightly higher average error on the test data.
The standard deviation of MSE has decreased from 25.95 to 15.30. A lower standard deviation indicates that the model's performance is more consistent across different runs or subsets of the data. In this case, the increased number of epochs seems to have contributed to a more stable performance.







